spec/lib/appsignal/rack/hanami_instrumentation_spec.rb

```diff
@@ -0,0 +1,111 @@
+require "appsignal/rack/hanami_instrumentation"
+
+describe Appsignal::Rack::HanamiInstrumentation do
+  class RecognizedRoute
+    def self.action
+      "mockcontroller#show"
+    end
+  end
+
+  class Routes
+    def self.recognize(_path)
+      RecognizedRoute
+    end
+  end
+
+  class Web
+    def self.routes
+      Routes
+    end
+  end
+
+  before :context do
+    start_agent
+  end
+
+  let(:app) { double(:call => true) }
+  let(:env) { { :path => "/", :method => "GET" } }
+  let(:options) { {} }
+  let(:middleware) { Appsignal::Rack::HanamiInstrumentation.new(app, options) }
+
+  describe "#call" do
+    before do
+      allow(middleware).to receive(:raw_payload).and_return({})
+    end
+
+    context "when appsignal is active" do
+      before { allow(Appsignal).to receive(:active?).and_return(true) }
+
+      it "should call with monitoring" do
+        expect(middleware).to receive(:call_with_appsignal_monitoring).with(env)
+      end
+    end
+
+    context "when appsignal is not active" do
+      before { allow(Appsignal).to receive(:active?).and_return(false) }
+
+      it "should not call with monitoring" do
+        expect(middleware).to_not receive(:call_with_appsignal_monitoring)
+      end
+
+      it "should call the stack" do
+        expect(app).to receive(:call).with(env)
+      end
+    end
+
+    after { middleware.call(env) }
+  end
+
+  describe "#call_with_appsignal_monitoring", :error => false do
+    it "should create a transaction" do
+      expect(Appsignal::Transaction).to receive(:create).with(
+        kind_of(String),
+        Appsignal::Transaction::HTTP_REQUEST,
+        kind_of(Rack::Request)
+      ).and_return(double(:set_action_if_nil => nil, :set_http_or_background_queue_start => nil, :set_metadata => nil))
+    end
+
+    it "should call the app" do
+      expect(app).to receive(:call).with(env)
+    end
+
+    context "with an error", :error => true do
+      let(:error) { VerySpecificError.new }
+      let(:app) do
+        double.tap do |d|
+          allow(d).to receive(:call).and_raise(error)
+        end
+      end
+
+      it "should set the error" do
+        expect_any_instance_of(Appsignal::Transaction).to receive(:set_error).with(error)
```

Comment:
Looks like you might have copied this from our Rack setup. We're actually in the process of rewriting these tests (when we have some spare time) to remove the amount of mocking and stubbing we have in our test suite.

We've added a new way of testing transaction data which allows us to test the same without `expect(transaction).to receive(:method)` for every call. The idea being: It shouldn't really matter how the data is set, as long as the transaction contains the data.

You can see how I applied this is in our ActionCable integration, which was added quite recently in PR #309.

```ruby
describe "some integration" do
  let(:env) { { :foo => :bar } }
  let(:transaction) do
    Appsignal::Transaction.new(
      SecureRandom.uuid,
      Appsignal::Transaction::HTTP_REQUEST,
      ActionDispatch::Request.new(env)
    )
  end
  before do
    expect(Appsignal::Transaction).to receive(:create)
      .with(transaction_id, Appsignal::Transaction::HTTP_REQUEST, kind_of(ActionDispatch::Request))
      .and_return(transaction)
    allow(Appsignal::Transaction).to receive(:current).and_return(transaction)
    # Make sure sample data is added
    expect(transaction.ext).to receive(:finish).and_return(true)
    # Stub complete call, stops it from being cleared in the extension
    # And allows us to call `#to_h` on it after it's been completed.
    expect(transaction.ext).to receive(:complete)
  end

  it "creates an AppSignal transaction" do
    # call code to perform action in integration

    expect(transaction.to_h).to include(
      "action" => "Hanami#action",
      "error" => nil,
      "id" => transaction.transaction_id,
      "namespace" => Appsignal::Transaction::HTTP_REQUEST,
      "metadata" => {
        "method" => "GET",
        "path" => "/blog"
      }
    )

    # same for `transaction.to_h["events"]` is the integration creates any
    # and same for
    # - `transaction.to_h["sample_data"]`
    # - `transaction.to_h["error"]`
  end
end
```

This lengthy comment is nothing specific to your PR, but we'd like to limit the amount of mocking and stubbing we do here. If we can test it through the full Ruby gem code and our extension like this we're a lot more sure everything works like it should.

It would be great if you can have a look at this new setup. Otherwise we'll be able to update it to the new setup, but since adding Hanami integration was unplanned I'm sure when exactly we'll get around to it so we'll put it on our backlog and discuss at the next backlog meeting (which is weekly).